keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1085 @@
+
+#### During Validation/Admission
+
+These are potential failures that could arise during validation or submission that the API server performs when sending the modified StatefulSet object. In the diagram above, it refers to the arrow denoted with `(1)`.  
+
+We won't expand on any of the errors that could happen during this phase right now, as we'll talk about it during [Validation](#validation)
+But it's worth noting that this is the ideal place that errors could happen in - as it'll give the user immediate feedback. 
+
+#### When Patching The PVC Object
+
+These are errors that could happen when the reconciliation logic in the StatefulSet controller attempts to perform a `patch` operation on the PVC object. These errors arise from either the PVC validation logic in the API server or its admission plugin. In the diagram above, it refers to the arrow denoted with `(6)`.
+
+* The storage class of the volume we're trying to resize doesn't allow for expansions.
+  * It shouldn't be a problem to have this validation in the StatefulSet validation/admission and we'll talk about it during [Validation](#validation).
+* We are trying to decrease the size of the volume.
+  * Specifically, we're trying to decrease the size of the volume to a size that's less (or equal) than it's capacity. Decreasing the size of the volume to a size that's greater than its current capacity is allowed, as its part of the recovery from failed expansions mechanism, described in [KEP 1790](../1790-recover-resize-failure/README.md).
+* The volume we're trying to resize is not bound.
+  * This shouldn't be an issue because as we'll discuss later, we'll apply to reconciliation logic to the PVCs only when the pod that they're bound to is healthy and running.
+
+#### Asynchronous Errors/Failures
+
+This refers to a class of errors that could happen after the StatefulSet controller has successfully performed a `patch` operation on the PVC object. These errors are asynchronous in nature and can happen at any time after the `patch` operation. In the diagram above, it refers to the arrow denoted with `(10)`. (Though in reality, `kubelet` is involved in the resize as well, when an online expansion is performed, but we leave it out of the diagram).
+
+* The CSI driver doesn't support online expansions.
+  * This is a scenario where the driver cannot do an expansion while the volume is published to a node/the pod that's using it is up. When this happens, the CSI driver will return an error when the CSI resizer calls it (eg `FailedPrecondition`). In this proposal we won't be supporting the scenario of not supporting online expansions. Ideally, we'd want to know ahead of time (during StatefulSet admission/validation) if the CSI driver supports online expansion or not - but it might not be so simple. We'll discuss it in details when we talk about [Validation](#validation).
+* The CSI driver doesn't support offline expansions.
```

Comment:
So this section says - "We will talk about in this validation section". But reading online capabilities  section, no solution is proposed.

Reply:
Do we intend to ever support offline expansion of statefulsets? Because supporting this will require changes in statefulset controller. The pod must only be recreated after volume has been expanded.

Reply:
> Do we intend to ever support offline expansion of statefulsets?

Not in this KEP but it can definetely be something that's addressed in a follow up